crates/firewheel-core/src/diff/mod.rs

```diff
@@ -0,0 +1,136 @@
+use crate::{
+    collector::ArcGc,
+    event::{NodeEventType, ParamData},
+};
+
+use smallvec::SmallVec;
+
+mod collections;
+mod leaf;
+
+pub use firewheel_macros::{Diff, Patch};
+
+/// A trait for diffing and patching.
+///
+/// This trait allows a type to perform diffing on itself,
+/// generating events that another instance can use to patch
+/// itself.
+///
+/// Fields are distinguished by their [`ParamPath`]. Since
+/// every non-sharing struct can be represented as a tree,
+/// a path of indeces can be used to distinguish any
+/// arbitrarily nested field.
+pub trait Diff {
+    /// Compare `self` to `baseline` and generate events to resolve any differences.
+    fn diff<E: EventQueue>(&self, baseline: &Self, path: PathBuilder, event_queue: &mut E);
+}
+
+pub trait Patch {
+    /// Patch `self` according to the incoming data.
+    /// This will generally be called from within
+    /// the audio thread.
+    ///
+    /// `data` is intentionally made a shared reference.
+    /// This should make accidental syscalls due to
+    /// additional allocations or drops more difficult.
+    /// If you find yourself reaching for interior
+    /// mutability, consider whether you're building
+    /// realtime-appropriate behavior.
+    fn patch(&mut self, data: &ParamData, path: &[u32]) -> Result<(), PatchError>;
+
+    /// Patch a set of parameters with incoming events.
+    ///
+    /// This is usefule as a convenience method for extracting the path
+    /// and data components from a top-level [`NodeEventType`].
+    /// It also allows types to use all of [`NodeEventType`]'s variants
+    /// for patching, if necessary.
+    fn patch_params(&mut self, event: &NodeEventType) {
+        if let NodeEventType::Param { data, path } = event {
+            // NOTE: It may not be ideal to ignore errors.
+            // Would it be possible to log these in debug mode?
+            let _ = self.patch(data, path);
+        }
+    }
+}
+
+/// A path of indeces that uniquely describes an arbitrarily nested field.
+pub enum ParamPath {
+    Single(u32),
+    Multi(ArcGc<SmallVec<[u32; 4]>>),
```

Comment:
If we're wrapping the param path in an `ArcGc` anyway, then I don't think there is any benefit to using SmallVec (in fact it would lead to a double indirection for longer paths). But correct me if I'm wrong on this.

Reply:
If we want to keep the overall event size down to 32 bytes, we can't avoid double indirection _somewhere_. A simple `ArcGc<[u32]>` actually bumps it up to 40 bytes because a slice is a wide pointer.

The `SmallVec` should limit double-indirection for most types.

It looks like I misplaced my comment on this, here's what is says just a few lines later:

```rs
// NOTE: Using a `SmallVec` instead of a `Box<[u32]>` yields
// around an 8% performance uplift for cases where the path
// is in the range 2..=4.
//
// Beyond this range, the performance drops off around 13%.
//
// Since this avoids extra allocations in the common < 5
// scenario, this seems like a reasonable trade off.
```

The performance delta was determined with the benchmarking included in this PR.

Reply:
Ah ok, yeah that makes sense.